### ISSUE TITLE
Improve startup time

### ISSUE THREAD

**Issue by bbondy** [brave/browser-laptop#3025]
Using `console.time` and `console.timeEnd` figure out why our startup time isn't ideal and find ways to improve it by lazily loading the slow things.  Or see what we can wait until a few seconds into the app is started to init.

**Comment by alexwykoff** [comment]
+1 from support : https://linkbubble.zendesk.com/agent/tickets/6457

**Comment by srirambv** [comment]
+1 from support: https://linkbubble.zendesk.com/agent/tickets/6707

**Comment by bbondy** [comment]
1 from support: https://linkbubble.zendesk.com/agent/tickets/6630

**Comment by bbondy** [comment]
+1 from support: https://linkbubble.zendesk.com/agent/tickets/6573

**Comment by cezaraugusto** [comment]
+1 from support: https://linkbubble.zendesk.com/agent/tickets/5622

**Comment by Liunkae** [comment]
+ 1 from #5677.

**Comment by Liunkae** [comment]
UI suggestion: #5820

**Comment by luixxiul** [comment]
FWIW: https://github.com/electron/electron/issues/5672

**Comment by bsclifton** [comment]
Also see https://github.com/atom/atom/issues/13253

We can share details with upstream in https://github.com/electron/maintainers-private/issues/32

**Comment by Renegilo** [comment]
There is  still the problem of Brave's slow startup that doesn't seem to have been fixed since milestone 1.1.0. Its becoming my favorite browser, I'm I'm finding myself opening Firefox or Edge while waiting for it to start. And by the time Brave starts the other mentioned browser has already open and reached the the site I needed. The current version I am now using is 
Name	Version
Brave	0.12.15
Muon	1.4.31
libchromiumcontent	53.0.2785.143
V8	5.3.332.47
Node.js	6.5.0
Update Channel	dev
os.platform	win32
os.release	10.0.14393
os.arch	x64

**Comment by luixxiul** [comment]
> have been fixed since milestone 1.1.0. 

Because it's the milestone in a future.

**Comment by aekeus** [comment]
Will instrument with https://github.com/brave/browser-laptop/pull/6267

**Comment by srirambv** [comment]
+1 from community https://community.brave.com/t/brave-is-the-slowest-to-load/2688

**Comment by Jacalz** [comment]
+1 from community
https://community.brave.com/t/faster-startup-times/2917/2

**Comment by bsclifton** [comment]
+1 from https://github.com/brave/browser-laptop/issues/9596

**Comment by Miyurz** [comment]
This is still an issue. Do we know which release is going to fix this ?

**Comment by bsclifton** [comment]
@Miyurz this is something we'd definitely love to improve.  Unfortunately, I can't give a date. You can track some of our performance issues here with https://github.com/brave/browser-laptop/issues/9311

**Comment by eljuno** [comment]
+1 [from community](https://community.brave.com/t/lastpass-doesnt-work-sites-automatically-logged-out-after-browser-reboot/5059/4?u=eljuno)

**Comment by NejcZdovc** [comment]
## Startup time research

### with a lot of bookmarks 75k, one window one tab (new tab), default settings

**Measurements done with reducers timers**

To see the skeleton of a browser (page is still not loaded) it takes **~12 seconds** to load
- [ ] ledger.js (`APP_SET_STATE`) adds ~300ms
- [ ] urlBarSuggestionsReducer.js (`APP_SET_STATE`) adds ~500ms
- [ ] menu.js (`APP_SET_STATE`) adds ~2200ms

**Measurements done with dev tool profiler**
<img width="1071" alt="bookmarks" src="https://user-images.githubusercontent.com/9574457/29524850-3db91066-8691-11e7-8c77-28ae69670b58.png">

### 15 windows with 10 new tabs on each, no bookmarks, default settings

**Measurements done with reducers timers**

To load all 15 windows and to see new tab page on each of it takes **+42 seconds** to load, **32 seconds** to only see url bar without tabs. Startup is not that bad, it only takes ~2 second to display white window.
- [ ] tabsReducer (`APP_SET_STATE`) add  ~**300ms**


**Measurements done with dev tool profiler**
<img width="1075" alt="windows" src="https://user-images.githubusercontent.com/9574457/29524865-464a14be-8691-11e7-8cb0-8724f5266b96.png">

**Comment by srirambv** [comment]
+1 from community https://community.brave.com/t/brave-browser-is-slow-when-opening/6718/2

**Comment by pjm77** [comment]
Yessss. Please make it a higher priority! People start flocking away from Chrome and Mozilla right now in relation to censorship and privacy issues. Brave is such an obvious choice. I've checked it out and it suits me just fine, but a 20 second startup just to get an empty browser is a no go for somebody like me and many others. If you start a browser once a day it's not a biggie, but if's 15 times a day it becomes a serious issue. It's the only thing that keeps me from using it as my primary browser.

**Comment by eljuno** [comment]
+1 from community https://community.brave.com/t/moved-back-to-chrome-for-now/7342?u=eljuno

**Comment by zentagonist** [comment]
Converted this issue to an Epic so we can start tracking specific issues inside this one.

**Comment by cezaraugusto** [comment]
Per @zentagonist request, I'm typing down what we could cover with this. The plan is to split this issue most we can so it's measurable. 

* Remove task-specific calls for unneeded features in start-up time (does ledger needs to be called at init time?)
* Lazy-load task-specific calls likely to be used first once the app is ready (such as suggestions and bookmarks)
* Improve the initial loading experience. Measure with metrics for first paint and first meaningful paint.
* If all else fails, i.e. a feature that could not be removed at first content load (like punycode?), reduce call cost for long task operations by replacing them with faster methods

@NejcZdovc please add notes here based on your research if there's something more we should cover. Plan is to split this into several sub-tasks so that's achievable and not stale based on many issues to be covered

**Comment by MacStasy** [comment]
Any update on the status of the fix?

**Comment by bsclifton** [comment]
Great fix by @ayumi which needs some additional work before we can accept it (if anyone is interested in checking out the hooks):
https://github.com/brave/browser-laptop/pull/11962

**Comment by srirambv** [comment]
+1 from https://community.brave.com/t/low-performance-brave-0-21-24-vs-chrome-65-0-3325-181/18379

**Comment by NumDeP** [comment]
+1

**Comment by srirambv** [comment]
+1 from https://community.brave.com/t/delay-on-start-up/28115

**Comment by srirambv** [comment]
This should be fixed with brave-core but still needs tracking on brave-core once brave specific features are implemented.

**Comment by srirambv** [comment]
+1 by @Xenios91  via https://github.com/brave/brave-browser/issues/620

**Comment by nisc** [comment]
Latest release from 14 days ago takes ca. 20 seconds to load on a modern Windows notebook with SSD. Chrome takes around 1 second on the same machine. Has this been fixed in brave-core yet?

> Brave: 0.23.79 
> V8: 6.8.275.24 
> rev: 51b49051a779f0db94fbcfd0df5faca781299ea0 
> Muon: 8.0.7 
> OS Release: 6.3.9600 
> Update Channel: Release 
> OS Architecture: x64 
> OS Platform: Microsoft Windows 
> Node.js: 7.9.0 
> Brave Sync: v1.4.2 
> libchromiumcontent: 68.0.3440.84

**Comment by bsclifton** [comment]
@nisc did you want to give it a shot? You can try one of our releases available here:
https://github.com/brave/brave-browser/releases

**Comment by nisc** [comment]
@bsclifton 0.54.1 reduces startup time by ~80% to ~4 seconds on my machine. Much better than the >20 seconds in the stable release. Still not as fast as Chrome. But fast enough to be useful in production. This can stay closed.
